### PR TITLE
HPCC-22806 Handle XML encoding &#nnn; properly in ESP

### PR DESCRIPTION
**Pull request by gaojx**
## Type of change:
- [x] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [ ] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [x] I have added tests to cover my changes.
  - [x] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [ ] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

### CODE REVIEW

**Comment by gaojx** [comment]
@mayx @afishbeck please review.

File: esp/bindings/SOAP/xpp/sxt/XmlTokenizer.h [hunk 1]
```diff
@@ -201,6 +201,56 @@ namespace sxt {
...
+    {
+      outBuf[idx++] = (unsigned char)(0xFC | c >> 30);
+      outBuf[idx++] = (unsigned char)(0x80 | c >> 24 & 0x3F);
+      outBuf[idx++] = (unsigned char)(0x80 | c >> 18 & 0x3F);
+      outBuf[idx++] = (unsigned char)(0x80 | c >> 12 & 0x3F);
+      outBuf[idx++] = (unsigned char)(0x80 | c >> 6 & 0x3F);
+      outBuf[idx++] = (unsigned char)(0x80 | c & 0x3F);
+    }
+    else
+		  assert(false);
```
Comment (richardkchapman) [inline esp/bindings/SOAP/xpp/sxt/XmlTokenizer.h]: throwUnexpected() would be better I think

File: esp/bindings/SOAP/xpp/sxt/XmlTokenizer.h [hunk 2]
```diff
@@ -201,6 +201,59 @@ namespace sxt {
...
+      outBuf[idx++] = (unsigned char)(0xFC | c >> 30);
+      outBuf[idx++] = (unsigned char)(0x80 | c >> 24 & 0x3F);
+      outBuf[idx++] = (unsigned char)(0x80 | c >> 18 & 0x3F);
+      outBuf[idx++] = (unsigned char)(0x80 | c >> 12 & 0x3F);
+      outBuf[idx++] = (unsigned char)(0x80 | c >> 6 & 0x3F);
+      outBuf[idx++] = (unsigned char)(0x80 | c & 0x3F);
+    }
+    else {
+      char buf[256];
+      snprintf(buf, 256, "Internal error at %s(%d) in %s", __FILE__, __LINE__, __func__);
```
Comment (afishbeck) [inline esp/bindings/SOAP/xpp/sxt/XmlTokenizer.h]: Could provide a bit more description.  Such as "invalid UCS character" or "out of bounds UCS character"...
 at %s(%d) in %s.

**Review by afishbeck — CHANGES_REQUESTED** [review]
@gaojx mostly changes to make things clearer.  Both in code and in error desc.

**Review by afishbeck — APPROVED** [review]
@gaojx looks good.

**Review by afishbeck — CHANGES_REQUESTED** [review]
@gaojx actually isn't clean, please squash and make sure it's a clean merge, I see unrelated changes.

File: roxie/udplib/udptopo.cpp [hunk 3]
```diff
@@ -142,7 +142,7 @@ CTopologyServer::CTopologyServer(const char *topologyInfo)
                 slaves[channel].append(ep);
                 if (ep.equals(mySlaveEP))
                 {
-                    mySubChannels[channel] = slaves[channel].ordinality();
+                    mySubChannels[channel] = slaves[channel].ordinality()-1;
```
Comment (afishbeck) [inline roxie/udplib/udptopo.cpp]: @gaojx Looks like this needs rebasing.

**Comment by gaojx** [comment]
It seems that rebase brought in more changes. This pull request is now very messy. I must did something wrong.  Would it be better for me to abandon the feature branch and restart a new branch from scratch?

**Comment by richardkchapman** [comment]
git rebase --onto candidate-7.6.0 HEAD^ should fix it. (followed by a force-push).

**Comment by gaojx** [comment]
I did: 
git rebase --onto candidate-7.6.0 HEAD^
git push origin HPCC22806-XML-entity-reference-not-handled-properly-in-ESP --force

It brought in one more change. What should we do now?

**Pull request closed without merging**